runtime/src/iree/io/formats/irpa/irpa_parser.c

```diff
@@ -0,0 +1,348 @@
+                            "metadata segment reference out of range (%" PRIu64
+                            " to %" PRIu64 " for %" PRIu64
+                            ", segment_size=%" PRIu64 ")",
+                            range.offset, range.offset + range.length - 1,
+                            range.length, header->metadata_segment.length);
+  }
+  iree_io_physical_offset_t span_offset =
+      base_offset + header->metadata_segment.offset + range.offset;
+  *out_span =
+      iree_make_const_byte_span(file_contents.data + span_offset, range.length);
+  return iree_ok_status();
+}
+
+static iree_status_t iree_io_resolve_irpa_v0_storage(
+    iree_const_byte_span_t file_contents, iree_io_physical_offset_t base_offset,
+    const iree_io_parameter_archive_header_v0_t* header,
+    iree_io_parameter_archive_storage_ref_t range,
+    iree_io_physical_offset_t* out_offset) {
+  *out_offset = 0;
+  if (range.length == 0) return iree_ok_status();
+  if (range.offset + range.length > header->storage_segment.length) {
+    return iree_make_status(IREE_STATUS_OUT_OF_RANGE,
+                            "storage segment reference out of range (%" PRIu64
+                            " to %" PRIu64 " for %" PRIu64
+                            ", segment_size=%" PRIu64 ")",
+                            range.offset, range.offset + range.length - 1,
+                            range.length, header->storage_segment.length);
+  }
+  iree_io_physical_offset_t storage_offset =
+      base_offset + header->storage_segment.offset + range.offset;
+  *out_offset = storage_offset;
+  return iree_ok_status();
+}
+
+static iree_status_t iree_io_parse_irpa_v0_splat_entry(
+    const iree_io_parameter_archive_header_v0_t* header,
+    const iree_io_parameter_archive_splat_entry_t* splat_entry,
+    iree_string_view_t name, iree_const_byte_span_t metadata,
+    iree_io_parameter_index_t* index) {
+  if (splat_entry->header.entry_size < sizeof(*splat_entry)) {
+    return iree_make_status(IREE_STATUS_OUT_OF_RANGE,
+                            "splat entry length underflow");
+  }
+  if (splat_entry->pattern_length > sizeof(splat_entry->pattern)) {
+    return iree_make_status(IREE_STATUS_INVALID_ARGUMENT,
+                            "splat pattern length %u out of bounds %" PRIhsz,
+                            splat_entry->pattern_length,
+                            sizeof(splat_entry->pattern));
+  }
+  iree_io_parameter_index_entry_t entry = {
+      .key = name,
+      .metadata = metadata,
+      .length = splat_entry->length,
+      .type = IREE_IO_PARAMETER_INDEX_ENTRY_STORAGE_TYPE_SPLAT,
+      .storage =
+          {
+              .splat =
+                  {
+                      // .pattern = splat_entry->pattern,  below
+                      .pattern_length = splat_entry->pattern_length,
+                  },
+          },
+  };
+  memcpy(entry.storage.splat.pattern, splat_entry->pattern,
+         entry.storage.splat.pattern_length);
```

Comment:
I'm a bit confused by the comment here. Seems like a struct member is uninitialized here? Maybe set to NULL/0 explicitly?

Reply:
added zero set, but the memcpy is needed - the input pattern pointer only has pattern_length valid bytes and we can't statically initialize it inline